File: src/main/java/dsa/fundamentals/bag.go
package main

import "fmt"

type Bag[C any] struct {
	elements []C
	iterator *BagIterator[C]
}

func NewStringBag() *Bag[string] {
	b := &Bag[string]{
		elements: make([]string, 0),
	}

	iterator := &BagIterator[string]{
		bag:   b,
		index: 0,
	}
	b.iterator = iterator
	return b
}

func (d *Bag[C]) Add(item C) {
	d.elements = append(d.elements, item)
}

func (d *Bag[C]) Size() int {
	return len(d.elements)
}

func (d *Bag[C]) IsEmpty() bool {
	return d.Size() == 0
}

func (d *Bag[C]) Iterator() *BagIterator[C] {
	return d.iterator
}

type BagIterator[C any] struct {
	index int
	bag   *Bag[C]
}

func (b *BagIterator[C]) HasNext() bool {
	return b.index < b.bag.Size()

}
func (b *BagIterator[C]) Next() C {
	if !b.HasNext() {
		panic("No elements")
	}
	el := b.bag.elements[b.index]
	b.index++
	return el
}

// ***************** Testing *****************
func main() {
	stringBag := NewStringBag()

	stringBag.Add("Item 1")
	stringBag.Add("Item 2")

	fmt.Println(stringBag.Size())

	iterator := stringBag.Iterator()

	for iterator.HasNext() {
		fmt.Println(iterator.Next())
	}
}

File: src/main/java/dsa/fundamentals/Bag.java
package dsa.fundamentals;

import java.util.Iterator;

public final class Bag<T extends Comparable<T>> implements Iterable<T> {

    private Object[] elements;
    private int capacity;
    private int size;

    public Bag() {
        this(12);
    }

    public Bag(int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException("capacity must be greater than 0");
        }
        this.capacity = capacity;
        elements = new Object[capacity];
    }

    public void add(T item) {
        ensureCapacity();
        elements[size++] = item;
    }

    private void ensureCapacity() {
        if (size == capacity) {
            this.capacity = capacity * 2;
            var temp = new Object[this.capacity];
            System.arraycopy(elements, 0, temp, 0, size);
            elements = temp;
        }
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iterator<>() {
            private int index = 0;

            @Override
            public boolean hasNext() {
                return Bag.this.size > index;
            }

            @Override
            public T next() {
                return (T) Bag.this.elements[index++];
            }
        };
    }
}
